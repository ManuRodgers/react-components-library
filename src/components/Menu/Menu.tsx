/* eslint-disable react/prop-types */
import classnames from 'classnames';
import React, {
  CSSProperties,
  FC,
  ReactNode,
  createContext,
  memo,
  useCallback,
  useState,
} from 'react';

type MenuMode = 'horizontal' | 'vertical';
type SelectCallback = (selectedIndex: string) => void;
export interface MenuProps {
  mode?: MenuMode;
  style?: CSSProperties;
  className?: string;
  defaultIndex?: string;
  onSelect?: SelectCallback;
  defaultOpenSubMenus?: string[];
  children?: ReactNode;
}

export interface MenuContext {
  index: string;
  onSelect?: SelectCallback;
  mode?: MenuMode;
  defaultOpenSubMenus?: string[];
}

export const ExportedMenuContext = createContext<MenuContext>({ index: '0' });

const Menu: FC<MenuProps> = ({
  defaultIndex,
  className,
  mode,
  onSelect,
  style,
  defaultOpenSubMenus,
  children,
}) => {
  const [currentActive, setCurrentActive] = useState<string>(
    defaultIndex || '0'
  );
  const classes = classnames('manu-menu', className, {
    'menu-horizontal': mode === 'horizontal',
    'menu-vertical': mode === 'vertical',
  });
  const handleClick = useCallback(
    (index: string) => {
      setCurrentActive(index);
      onSelect && onSelect(index);
    },
    [onSelect]
  );
  const passedContext: MenuContext = {
    index: currentActive ? currentActive : '0',
    onSelect: handleClick,
    defaultOpenSubMenus,
    mode,
  };
  return (
    <ul className={classes} style={style} data-testid={'test-menu'}>
      <ExportedMenuContext.Provider value={passedContext}>
        {children}
      </ExportedMenuContext.Provider>
    </ul>
  );
};
Menu.defaultProps = {
  defaultIndex: '0',
  mode: 'horizontal',
  defaultOpenSubMenus: [],
};

export default memo(Menu);
