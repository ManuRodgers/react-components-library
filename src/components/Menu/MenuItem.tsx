import classnames from 'classnames';
import React, { CSSProperties, FC, memo, useCallback, useContext } from 'react';

import { ExportedMenuContext } from './Menu';

export interface MenuItemProps {
  index?: string;
  disabled?: boolean;
  className?: string;
  style?: CSSProperties;
  children: React.ReactNode;
}

const MenuItem: FC<MenuItemProps> = ({
  className,
  style,
  children,
  disabled,
  index,
}: MenuItemProps): JSX.Element => {
  const context = useContext(ExportedMenuContext);
  const classes = classnames('menu-item', className, {
    'is-disabled': disabled,
    'is-active': context.index === index,
  });
  const handleClick = useCallback(() => {
    if (context.onSelect && !disabled && index) {
      context.onSelect(index);
    }
  }, [context, disabled, index]);

  return (
    <li style={style} className={classes} onClick={handleClick}>
      {children}
    </li>
  );
};
MenuItem.displayName = 'MenuItem';
export default memo(MenuItem);
