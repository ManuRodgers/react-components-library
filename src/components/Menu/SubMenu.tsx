import classnames from 'classnames';
import React, {
  FunctionComponent,
  FunctionComponentElement,
  ReactNode,
  memo,
  useCallback,
  useContext,
  useMemo,
  useState,
} from 'react';

import Icon from '../Icon/Icon';
import Transition from '../Transition/Transition';
import { ExportedMenuContext } from './Menu';
import { MenuItemProps } from './MenuItem';

interface SubMenuProps {
  title: string;
  index?: string;
  className?: string;
  style?: React.CSSProperties;
  children: ReactNode;
}

const SubMenu: FunctionComponent<SubMenuProps> = ({
  children,
  className,
  style,
  title,
  index,
}) => {
  const context = useContext(ExportedMenuContext);
  const openedSubMenus = context.defaultOpenSubMenus as string[];
  const isOpened =
    index && context.mode === 'vertical'
      ? openedSubMenus.includes(index)
      : false;
  const [openSubMenu, setOpenSubMenu] = useState<boolean>(isOpened);
  const classes = classnames('menu-item submenu-item', className, {
    'is-active': context.index === index,
    'is-opened': openSubMenu,
    'is-vertical': context.mode === 'vertical',
  });

  const renderChildren = useCallback(() => {
    const subMenuClasses = classnames('manu-submenu', {
      'menu-opened': openSubMenu,
    });
    const childrenComponent = React.Children.map(children, (child, i) => {
      const childElement = child as FunctionComponentElement<MenuItemProps>;
      return React.cloneElement(childElement, {
        index: `${index}-${i}`,
      });
    });

    return (
      <Transition in={openSubMenu} timeout={300} animation={'zoom-in-bottom'}>
        <ul className={subMenuClasses}>{childrenComponent}</ul>
      </Transition>
    );
  }, [children, index, openSubMenu]);

  const handleClick = useCallback(
    (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
      event.preventDefault();
      setOpenSubMenu(!openSubMenu);
    },
    [openSubMenu]
  );
  let timer: any;
  const handleMouse = useCallback(
    (event: React.MouseEvent<HTMLLIElement, MouseEvent>, toggle: boolean) => {
      clearTimeout(timer);
      event.preventDefault();
      timer = setTimeout(() => {
        setOpenSubMenu(toggle);
      }, 30);
    },
    [timer]
  );
  const clickEvents = useMemo(
    () => (context.mode === 'vertical' ? { onClick: handleClick } : {}),
    [context.mode, handleClick]
  );

  const mouseEvents = useMemo(() => {
    return context.mode === 'horizontal'
      ? {
          onMouseEnter: (
            event: React.MouseEvent<HTMLLIElement, MouseEvent>
          ): void => {
            handleMouse(event, true);
          },
          onMouseLeave: (
            event: React.MouseEvent<HTMLLIElement, MouseEvent>
          ): void => {
            handleMouse(event, false);
          },
        }
      : {};
  }, [context.mode, handleMouse]);

  return (
    <li {...mouseEvents} key={index} className={classes} style={style}>
      <div className="submenu-title" {...clickEvents}>
        {title}
        <Icon icon="angle-down" className="arrow-icon" />
      </div>
      {renderChildren()}
    </li>
  );
};
SubMenu.displayName = 'SubMenu';
export default memo(SubMenu);
