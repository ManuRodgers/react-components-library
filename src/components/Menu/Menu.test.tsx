import {
  RenderResult,
  cleanup,
  fireEvent,
  render,
} from '@testing-library/react';
import React from 'react';

import Menu, { MenuProps } from './Menu';
import MenuItem from './MenuItem';

const defaultProps: MenuProps = {
  defaultIndex: '0',
  onSelect: jest.fn(),
  className: 'test',
};
const verticalProps: MenuProps = {
  defaultIndex: '0',
  mode: 'vertical',
};
let wrapper: RenderResult,
  wrapper2: RenderResult,
  menuElement: HTMLElement,
  activeElement: HTMLElement,
  disabledElement: HTMLElement,
  xyz: HTMLElement,
  manu: HTMLElement;
function renderMenu(props: Partial<MenuProps> = {}): RenderResult {
  return render(
    <Menu {...defaultProps} {...props}>
      <MenuItem index={'0'}>active</MenuItem>
      <MenuItem index={'1'} disabled>
        disabled
      </MenuItem>
      <MenuItem index={'2'}>xyz</MenuItem>
      <MenuItem index={'3'}>manu</MenuItem>
    </Menu>
  );
}
describe('test Menu and MenuItem component in default(horizontal) mode', function () {
  beforeEach(async () => {
    // finish all of assignments
    wrapper = renderMenu();
    menuElement = await wrapper.findByTestId('test-menu');
    activeElement = await wrapper.findByText('active');
    disabledElement = await wrapper.findByText('disabled');
    xyz = await wrapper.findByText('xyz');
    manu = await wrapper.findByText('manu');
  });
  afterEach(cleanup);
  it('should render correct Menu and MenuItem based on default props', () => {
    expect(menuElement).toBeInTheDocument();
    expect(menuElement).toHaveClass('manu-menu test');
    expect(menuElement.querySelectorAll('li').length).toEqual(4);
    expect(activeElement).toHaveClass('menu-item is-active');
    expect(disabledElement).toHaveClass('menu-item is-disabled');
  });

  it('click items should change active and call the right callback', () => {
    fireEvent.click(xyz);
    expect(xyz).toHaveClass('is-active');
    expect(activeElement).not.toHaveClass('is-active');
    expect(defaultProps.onSelect).toHaveBeenCalledWith('2');
    fireEvent.click(disabledElement);
    expect(disabledElement).not.toHaveClass('is-active');
    expect(defaultProps.onSelect).not.toHaveBeenCalledWith('1');
    fireEvent.click(manu);
    expect(manu).toHaveClass('is-active');
    expect(defaultProps.onSelect).toHaveBeenCalledWith('3');
    expect(xyz).not.toHaveClass('is-active');
    expect(defaultProps.onSelect).not.toHaveBeenCalledWith('1');
  });

  it('matches snapshot', async () => {
    const { asFragment } = renderMenu();
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('test Menu and MenuItem component in vertical mode', () => {
  beforeEach(() => {
    wrapper2 = renderMenu(verticalProps);
  });
  afterEach(cleanup);
  it('should render vertical mode when mode is set to vertical', () => {
    const menuElement = wrapper2.getByTestId('test-menu');
    expect(menuElement).toHaveClass('menu-vertical');
  });

  it('matches snapshot', async () => {
    const { asFragment } = renderMenu(verticalProps);
    expect(asFragment()).toMatchSnapshot();
  });
});
