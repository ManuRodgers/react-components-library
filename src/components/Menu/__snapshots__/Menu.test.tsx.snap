// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Menu and MenuItem component in default(horizontal) mode matches snapshot 1`] = `
<DocumentFragment>
  <ul
    class="manu-menu test menu-horizontal"
    data-testid="test-menu"
  >
    <li
      class="menu-item is-active"
    >
      active
    </li>
    <li
      class="menu-item is-disabled"
    >
      disabled
    </li>
    <li
      class="menu-item"
    >
      xyz
    </li>
    <li
      class="menu-item"
    >
      manu
    </li>
  </ul>
</DocumentFragment>
`;

exports[`test Menu and MenuItem component in vertical mode matches snapshot 1`] = `
<DocumentFragment>
  <ul
    class="manu-menu test menu-vertical"
    data-testid="test-menu"
  >
    <li
      class="menu-item is-active"
    >
      active
    </li>
    <li
      class="menu-item is-disabled"
    >
      disabled
    </li>
    <li
      class="menu-item"
    >
      xyz
    </li>
    <li
      class="menu-item"
    >
      manu
    </li>
  </ul>
</DocumentFragment>
`;
